flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/metadata/FlinkMetadata.java

```diff
@@ -238,4 +239,20 @@ ImmutableBitSet getUniqueGroups(
             RelWindowProperties getWindowProperties(RelNode r, RelMetadataQuery mq);
         }
     }
+
+    /** Metadata about which combinations of columns are change log upsert identifiers. */
+    public interface ChangeLogUpsertKeys extends Metadata {
+        Method METHOD = Types.lookupMethod(ChangeLogUpsertKeys.class, "getChangeLogUpsertKeys");
+
+        MetadataDef<ChangeLogUpsertKeys> DEF =
+                MetadataDef.of(
+                        ChangeLogUpsertKeys.class, ChangeLogUpsertKeys.Handler.class, METHOD);
+
+        Set<ImmutableBitSet> getChangeLogUpsertKeys();
```

Comment:
I down think, we can have multi changelog keys  in a RelNode

Reply:
It is same to unique keys.
Consider rank, keys can be input unique keys and partitionKey + rank_number.

Reply:
In my understanding from the description of the issue, the new added key is used to describe how the change log stream is distributed in parallel, so the key can only be single one, correct me if I am wrong. 

on the other hand, Consider: changelog source + rank(top 1) + upsert sink, if both input unique keys and partitionKey + rank_number are changelog upsert keys,  how can we get the right decision as: when pk of sink is partition key we don't need any extra process before write to sink, but when pk of sink is the pk of source, we will need to add extra process before sink to take records disorder into consider.

Reply:
You can take a look to `getChangeLogUpsertKeys(rel: Rank, mq: RelMetadataQuery)`.

- If the pk of changelog source is an extra field (f1) not partition key(f0). OK, this rank operator not output UpsertKey f1, but output UpsertKey f0 + rank_number. So if upsert sink pk is f1, OK, we need resolve records disorder. If upsert sink pk is f0 + rank_number, it is OK, we don't need do anything.
- If the pk of changelog source is f0 + f2. Because this pk contains partition key, this rank operator can output UpsertKey source pk. Whether PK of sink is f0 + f2 or f0 + rank_number, we do not need do anything.

Reply:
for the point 2, if the upsert key is f0(just the same the distribution of the stream), hash(f0) can satisfy require of sink both (f0, f2) and (f0, rank number),  is it more intuitive?

Reply:
I understand that your requirement is to ensure the order when it is not forward, but the problem here is different.

Even if we know the distribution keys, the operator in the middle may also lead to the disorder of upsert keys, which we also need to identify.